tests/test_notebooks.py

```diff
@@ -0,0 +1,14 @@
+import glob
+import pytest
+import nbformat
+import papermill as pm
+
+
+all_notebooks = [n for n in glob.glob("tutorials/**/*.ipynb", recursive=True)]
+
+
+@pytest.mark.parametrize("notebook", sorted(all_notebooks))
+def test_notebooks(notebook):
+    """Test Notebooks in the tutorial root folder."""
```

Comment:
Would you please explain a bit more what testing here is supposed to test exactly. Like is it just to make sure no errors are present while running cells? For instance: "This test should pass if ...."

Reply:
This test run every notebook as part of CI so that we can catch any issues that popup in those tutorials asap. Isn;t this COOL?

For instance, if I add a cell with this invalid import
```python
import babeloud
```
I will get a failure like this
```
================================================================= FAILURES =================================================================
______________________________ test_notebooks[tutorials/Part 1 - (Getting Started) Remote Tokenization.ipynb] ______________________________

notebook = 'tutorials/Part 1 - (Getting Started) Remote Tokenization.ipynb'

    @pytest.mark.parametrize("notebook", sorted(all_notebooks))
    def test_notebooks(notebook):
        """Test Notebooks in the tutorial root folder."""
>       res = pm.execute_notebook(notebook, "/dev/null", parameters={}, timeout=300,)

tests/test_notebooks.py:13: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/share/c3/conda/envs/env3.6/lib/python3.6/site-packages/papermill/execute.py:108: in execute_notebook
    raise_for_execution_errors(nb, output_path)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

nb = {'cells': [{'cell_type': 'code', 'metadata': {'inputHidden': True, 'hide_input': True}, 'execution_count': None, 'sour...end_time': '2020-02-20T01:14:37.329375', 'duration': 1.328065, 'exception': True}}, 'nbformat': 4, 'nbformat_minor': 2}
output_path = '/dev/null'

    def raise_for_execution_errors(nb, output_path):
        """Assigned parameters into the appropriate place in the input notebook
    
        Parameters
        ----------
        nb : NotebookNode
           Executable notebook object
        output_path : str
           Path to write executed notebook
        """
        error = None
        for cell in nb.cells:
            if cell.get("outputs") is None:
                continue
    
            for output in cell.outputs:
                if output.output_type == "error":
                    if output.ename == "SystemExit" and (output.evalue == "" or output.evalue == "0"):
                        continue
                    error = PapermillExecutionError(
                        exec_count=cell.execution_count,
                        source=cell.source,
                        ename=output.ename,
                        evalue=output.evalue,
                        traceback=output.traceback,
                    )
                    break
    
        if error:
            # Write notebook back out with the Error Message at the top of the Notebook.
            error_msg = ERROR_MESSAGE_TEMPLATE % str(error.exec_count)
            error_msg_cell = nbformat.v4.new_code_cell(
                source="%%html\n" + error_msg,
                outputs=[
                    nbformat.v4.new_output(output_type="display_data", data={"text/html": error_msg})
                ],
                metadata={"inputHidden": True, "hide_input": True},
            )
            nb.cells = [error_msg_cell] + nb.cells
            write_ipynb(nb, output_path)
>           raise error
E           papermill.exceptions.PapermillExecutionError: 
E           ---------------------------------------------------------------------------
E           Exception encountered at "In [1]":
E           ---------------------------------------------------------------------------
E           ModuleNotFoundError                       Traceback (most recent call last)
E           <ipython-input-1-eb6323a4fbfd> in <module>
E                 1 # Hide warnings (nothing to do with SyferText)
E                 2 import warnings
E           ----> 3 import babeloud
E                 4 warnings.filterwarnings('ignore')
E           
E           ModuleNotFoundError: No module named 'babeloud'

/usr/local/share/c3/conda/envs/env3.6/lib/python3.6/site-packages/papermill/execute.py:194: PapermillExecutionError
```

Reply:
Thank you @dzlab for the example that is so cool. 

I actually meant that you add this to the test function docstring. Like just a simple statement telling that here we are just testing that tutorial jupyter notebooks run without errors. But there is no fine-grained testing with edge cases for instance.

Reply:
yeah that's this test goal, and we need one like this anyway to make sure changes are not breaking tutorials.